### ISSUE TITLE
presto worker node crashed at presto 0.160

### ISSUE THREAD

**Issue by wyukawa** [prestodb/presto#6916]
Hi

I use presto 0.160 + jdk1.8.0_101 + HDP 2.1 + hive connector + CentOS 6.5.

Suddenly, presto worker node crashed.
It's difficult to reproduce the problem.

presto query
```
SELECT
  a,b,c,d
  count(1) as ...,
  approx_distinct(...) as ...,
  approx_distinct(...) as ...
FROM
  ...
WHERE
  ...
AND
  yyyymmdd BETWEEN '20161201' AND '20161212'
GROUP BY CUBE (a,b,c,c)
```

presto log
```
com.facebook.presto.operator.PageTransportTimeoutException: Encountered too many errors talking to a worker node. The node may have crashed or be under too much load. This is probably a transient issue, so please retry your query in a few minutes. (http://host:port/v1/task/20161219_022327_01165_3mjz5.1.21/results/0/0 - requests failed for 60.00s)
	at com.facebook.presto.operator.HttpPageBufferClient$1.onFailure(HttpPageBufferClient.java:376)
	at com.google.common.util.concurrent.Futures$6.run(Futures.java:1310)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.RuntimeException: java.util.concurrent.TimeoutException: Idle timeout 30000 ms
	at com.google.common.base.Throwables.propagate(Throwables.java:160)
	at io.airlift.http.client.ResponseHandlerUtils.propagate(ResponseHandlerUtils.java:22)
	at com.facebook.presto.operator.HttpPageBufferClient$PageResponseHandler.handleException(HttpPageBufferClient.java:541)
	at com.facebook.presto.operator.HttpPageBufferClient$PageResponseHandler.handleException(HttpPageBufferClient.java:528)
	at io.airlift.http.client.jetty.JettyHttpClient$JettyResponseFuture.failed(JettyHttpClient.java:870)
	at io.airlift.http.client.jetty.JettyHttpClient$BufferingResponseListener.onComplete(JettyHttpClient.java:1104)
	at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:193)
	at org.eclipse.jetty.client.ResponseNotifier.notifyComplete(ResponseNotifier.java:185)
	at org.eclipse.jetty.client.HttpReceiver.terminateResponse(HttpReceiver.java:457)
	at org.eclipse.jetty.client.HttpReceiver.abort(HttpReceiver.java:528)
	at org.eclipse.jetty.client.HttpChannel.abortResponse(HttpChannel.java:129)
	at org.eclipse.jetty.client.HttpChannel.abort(HttpChannel.java:122)
	at org.eclipse.jetty.client.HttpExchange.abort(HttpExchange.java:257)
	at org.eclipse.jetty.client.HttpConversation.abort(HttpConversation.java:141)
	at org.eclipse.jetty.client.HttpRequest.abort(HttpRequest.java:708)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.abort(HttpConnectionOverHTTP.java:158)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.close(HttpConnectionOverHTTP.java:144)
	at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onIdleExpired(HttpConnectionOverHTTP.java:105)
	at org.eclipse.jetty.io.AbstractEndPoint.onIdleExpired(AbstractEndPoint.java:161)
	at org.eclipse.jetty.io.IdleTimeout.checkIdleTimeout(IdleTimeout.java:166)
	at org.eclipse.jetty.io.IdleTimeout$1.run(IdleTimeout.java:50)
	... 7 more
Caused by: java.util.concurrent.TimeoutException: Idle timeout 30000 ms
	... 11 more
```

hs_err_pid.log
https://gist.github.com/wyukawa/e3ae28ad332b3ae9ec2bf7947a47ffd6
https://gist.github.com/wyukawa/cce22932fd1384bd4122fd5d3f98d89d

Any suggestion?

**Comment by raghavsethi** [comment]
This usually happens when there is a host-level problem - stall/reboot, or there is a long full GC. Please let us know if this happens on the latest version, and if it can be reproduced.

**Comment by chengat1314** [comment]
hi @raghavsethi 
Thx for your clarification. 
Something has happened in our production cluster. 
And it is because of a long full GC.  (the worker is long GC will lost from the cluster)
Would you mind to give me some suggestions how to avoid that?
The long full GC was caused by a very big bad query. 
And meanwhile, all the concurrent queries failed after that.

Is there a way we can just let the big bad query fail and let the concurrent query continuous running  ?

FYI, our setting
http-server.threads.max=500
discovery-server.enabled=true
sink.max-buffer-size=1GB
query.max-memory=300GB
query.max-memory-per-node=15GB
query.max-history=300
query.min-expire-age=30m
http-server.http.port=8889
http-server.log.path=/var/log/presto/http-request.log
http-server.log.max-size=67108864B
http-server.log.max-history=5
log.max-size=268435456B
log.max-history=5
query.max-run-time = 60m
task.concurrency = 32
node-scheduler.network-topology = flat
exchange.http-client.request-timeout = 1200s
task.cpu-timer-enabled = false

using g1gc refer https://prestodb.io/docs/current/installation/deployment.html

Thank you very much!
Cheng Feng

**Comment by raghavsethi** [comment]
We've made a bunch of fixes that prevent full GCs since 0.160, especially in the last couple of releases. We will continue to make improvements, but please upgrade to the latest version and file a new issue with the query shapes that cause issues and GC log information.
